### ISSUE TITLE
Design Rollback strategy

### ISSUE THREAD

**Issue by AdheipSingh** [druid-io/druid-operator#81]
### MOTIVATION

- In order to achieve auto healing cluster, we should have an effective rollback strategy.
- Most common scenarios are production clusters should be able to rollback to previous stable state.
- As of now rolling upgrades can halt the upgrade when any node crashes, at this point if lets say if their no manual intervention for 5 min or pod restarts are more than 5 , the operator should rollback to previous version.

### IMPLEMENTATION

- Define state, define a version associated with each state. In operator terms the each apply of CR( kubectl apply -f cr.yaml ) we should  be map it with a version.
- As per k8s system, all manifests are stored in etcd where every CR has a resource version no which changes on every apply, this can be used to retrieve the previous manifests, but this a very complex/not so feasible approach as we might need to build separate etcd watchers etc.
- What i think can be achievable is that each manifest we apply, the operator can create a kubernetes secret encoding the entire CR, and name it has druid-<name>.<release-number>
- We can have an implementation which can decode the secret and reapply the configurations in case any of the above describe scenarios come up. 
- We can keep this as optional as some user might really want to see what caused the errors.

Just want to start an exploratory discussion, which can lead us to some concrete design.

**Comment by AdheipSingh** [comment]
@himanshug 
in case you missed this :) would like to get your thoughts on this one

**Comment by himanshug** [comment]
another issue: if a partial upgrade happened, then that might have updated some of the persisted state e.g. stuff stored in DB or segments that can only be understood by new code ... Druid code changes try to be backward compatible but don't care about forward compatibility... so it is possible that even if you reverted to a `last-known-good-version` , there is no guarantee that going back to that version would fix the cluster.

generally speaking, there would be a small subset among all of cluster upgrade problems which would be simply solved by changing the druid spec to a previous version. I am not sure if we can detect which cases those are, and blind revert may help or hurt... there might be a tradeoff to make.
the fault might not be in druid deployment at all, say coincidentally, k8s itself is having bad time to provision resource for whatever reason .. that should not inadvertently make operator take unnecessary actions.

> As per k8s system, all manifests are stored in etcd where every CR has a resource version no which changes on every apply, this can be used to retrieve the previous manifests, but this a very complex/not so feasible approach as we might need to build separate etcd watchers etc.

IMO it would not be any more complex than other ways proposed and we would just try to use versioned snapshots that k8s is already maintaining for all the resources including druid spec.

**Comment by AdheipSingh** [comment]
yes exactly, somehow its not possible to judge which state was the best state. 

My only cases were specific for prod cluster use cases, where if a faulty upgrades happens and their is no administrator to revert or debug, the operator automatically rollbacks to previous state. 

I also feel that ideally operator should emit an event to message queue and another service should kind of listen to the event and rollback the previous manifest.I believe this is possible and helm rollback can be leverage since it already is maintain g versioned manifests.  operator should not fiddle around the desired state, its sole purpose should just to  change the state of is desired.
